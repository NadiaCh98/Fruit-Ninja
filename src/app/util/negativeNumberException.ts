export class NegativeNumberException extends Error {
  constructor() {
    super('Value can not be negative');
  }
}
