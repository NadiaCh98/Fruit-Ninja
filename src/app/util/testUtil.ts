import { FruitSequence } from './../game/models/fruitData';
import { FruitsGenerator } from '../game/services/fruitsGenerator';

export const getFruitsSeriesByGeneratorAndIteraction = (
  generator: FruitsGenerator,
  iteraction: number
): FruitSequence => {
  return generator.generateFruitsSequence(iteraction);
};
