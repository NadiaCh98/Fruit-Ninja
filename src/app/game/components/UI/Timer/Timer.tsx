import React, { memo } from 'react';
import styled from 'styled-components';
import { RIGHT_POSITION } from '../../../../common/constant';
import { NegativeNumberException } from '../../../../util/negativeNumberException';
import { GameControls } from '../../GameControls/GameControls';

interface StyledTimerProps {
  readonly minutes: number;
  readonly seconds: number;
}

const StyledTimer = styled.p`
  font-size: var(--text-size);
  background: ${({ seconds, minutes }: StyledTimerProps) =>
    `var(--${minutes === 0 && seconds <= 10 ? 'red' : 'yellow'}-gradient)`};
  -webkit-background-clip: text;
  -webkit-text-fill-color: transparent;
`;

interface TimerProps {
  readonly time: number;
}

const convertMillisecondsToMinutes = (time: number): [number, number] => {
  const minutes = Math.floor(time / 60000);
  const seconds = (time % 60000) / 1000;
  return [minutes, seconds];
};

export const Timer: React.FC<TimerProps> = memo(({ time }) => {
  if (time < 0) {
    throw new NegativeNumberException();
  }
  const [minutes, seconds] = convertMillisecondsToMinutes(time);
  const timer = `${minutes}:${seconds < 10 ? `0${seconds}` : seconds}`;
  return (
    <GameControls style={RIGHT_POSITION}>
      <StyledTimer seconds={seconds} minutes={minutes}>
        {timer}
      </StyledTimer>
    </GameControls>
  );
});

Timer.displayName = 'Timer';
