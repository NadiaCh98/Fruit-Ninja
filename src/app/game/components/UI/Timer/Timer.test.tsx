import React from 'react';
import { render } from '@testing-library/react';
import { Timer } from './Timer';
import { NegativeNumberException } from '../../../../util/negativeNumberException';

describe('Timer', () => {
  it('has a correct template #1', () => {
    const { container } = render(<Timer time={90000} />);

    expect(container).toMatchSnapshot();
  });

  it('has a correct template #2', () => {
    const { container } = render(<Timer time={7000} />);

    expect(container).toMatchSnapshot();
  });

  it('asserts that Timer throws error if time < 0', () => {
    expect(() => render(<Timer time={-80} />)).toThrowError(
      NegativeNumberException
    );
  });
});
