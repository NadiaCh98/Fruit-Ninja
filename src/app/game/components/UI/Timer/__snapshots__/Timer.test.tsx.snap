// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Timer has a correct template #1 1`] = `
<div>
  <div
    class="sc-bdnxRM lhbDim"
    style="top: 0px; right: 0px;"
  >
    <p
      class="sc-gtsrHT ljxMPJ"
    >
      1:30
    </p>
  </div>
</div>
`;

exports[`Timer has a correct template #2 1`] = `
<div>
  <div
    class="sc-bdnxRM lhbDim"
    style="top: 0px; right: 0px;"
  >
    <p
      class="sc-gtsrHT kYMBSG"
    >
      0:07
    </p>
  </div>
</div>
`;
